django/db/models/base.py

```diff
@@ -1926,12 +1947,65 @@ def _check_constraints(cls, databases):
                         id='models.W038',
                     )
                 )
-            fields = (
-                field
-                for constraint in cls._meta.constraints if isinstance(constraint, UniqueConstraint)
-                for field in constraint.fields
-            )
+            fields = set()
+            references = set()
+            for constraint in cls._meta.constraints:
+                if isinstance(constraint, UniqueConstraint):
+                    fields.update(constraint.fields)
+                    if isinstance(constraint.condition, Q):
+                        references.update(cls._get_expr_references(constraint.condition))
+                elif isinstance(constraint, CheckConstraint) and isinstance(constraint.check, Q):
+                    references.update(cls._get_expr_references(constraint.check))
             errors.extend(cls._check_local_fields(fields, 'constraints'))
+            for field_name, *lookups in references:
+                try:
+                    field = cls._meta.get_field(field_name)
+                except FieldDoesNotExist:
+                    errors.append(
+                        checks.Error(
+                            "'constraints' refers to the nonexistent field '%s'." % field_name,
+                            obj=cls,
+                            id='models.E012',
+                        )
+                    )
+                    continue
+                if not field.is_relation:
+                    # Only related fields can result in joins.
+                    continue
+                if field.many_to_many:
+                    errors.append(
+                        checks.Error(
+                            "'constraints' refers to a ManyToManyField '%s', but "
+                            "ManyToManyFields are not permitted in 'constraints'." % field_name,
+                            obj=cls,
+                            id='models.E013',
+                        )
+                    )
+                    continue
+                if field.one_to_many:
+                    errors.append(
+                        checks.Error(
+                            "'constraints' refers to a reverse ForeignKey '%s', but "
+                            "reverse ForeignKey are not permitted in 'constraints'." % field_name,
+                            obj=cls,
+                            id='models.E042',
+                        )
+                    )
+                    continue
```

Comment:
This duplicates logic from `_check_local_fields()` and added unnecessary error `models.E042` which is already covered by `models.E012` in `_check_local_fields()`. I think we should pass fields from `references` to the `_check_local_fields()` and remove redundant logic, e.g.
```python
for field_name, *lookups in references:
    fields.add(field_name)
    if not lookups:
        # If it has no lookups it cannot result in a JOIN.
        continue
    try:
        field = cls._meta.get_field(field_name)
        if not field.is_relation or field.many_to_many or field.one_to_many:
            continue
    except FieldDoesNotExist:
        continue
    # JOIN must happen at the first lookup.
    first_lookup = lookups[0]
    if field.get_transform(first_lookup) is None and field.get_lookup(first_lookup) is None:
        errors.append(
            checks.Error(
                "'constraints' refers to '%s' which results a JOIN attempt, "
                "JOIN is not permitted in 'constraints'." % LOOKUP_SEP.join([field_name] + lookups),
                obj=cls,
                id='models.E042',
            )
        )
errors.extend(cls._check_local_fields(fields, 'constraints'))
```